### PR TITLE
Parse a users prayer list (pdf) and create new residents

### PR DESCRIPTION
**Pull request by ittullos**
This commit creates a new route /load_residents that will parse a pdf file and create new residents based on the contents of the pdf. The pdf file is expected to be a list of names and addresses. I have to do some odd logic to account for empty lines or lines that do not contain prayer names. I'm pretty sure that I am testing the incorrect thing here because the pdf is attached as a param. The code does sort the pdf and create the new residents. I just need to change how that pdf is passed with the request. I'm also having an issue getting my front end to make this call without getting a 401 error. I've been using the same headers that I use for the other requests, so kinda weird.

### CODE REVIEW

File: app/app.rb [hunk 1]
```diff
@@ -37,6 +37,15 @@
   devotionals.to_json
 end
 
+post '/prayer_list' do
```
Comment (zbarnard1) [inline app/app.rb:40]: This route is for `/prayer_list` but it doesn't create a PrayerList record, so it's a little bit surprising. Would this make more sense as a POST to `/user/residents`, instead?

File: app/models/user.rb [hunk 2]
```diff
@@ -1,8 +1,40 @@
...
 
+require 'pdf-reader'
+
 class User < ActiveRecord::Base
   validates_presence_of :first_name, :last_name, :email
   validates_uniqueness_of :email
 
   has_many :residents, -> { order(position: :asc) }
+
+  def load_residents(file)
```
Comment (zbarnard1) [inline app/models/user.rb:11]: I think this violates the Single Responsibility Principle, because now the User model has all these responsibilities related to parsing a file. It would be better to extract this code into a dedicated class.
I probably wouldn't even define a method on the User model. Once we have the `user_from_token` I would pass it to the class along with the file and let it do the work from there.

File: spec/requests/app_spec.rb [hunk 3]
```diff
@@ -46,6 +46,18 @@
     end
   end
 
+  describe 'POST /prayer_list' do
+    let(:pdf_file) { Rack::Test::UploadedFile.new('spec/fixtures/sample.pdf', 'application/pdf') }
```
Comment (zbarnard1) [inline spec/requests/app_spec.rb:50]: This is the downside of `git add -p`, you can sometimes miss new files.

File: app/models/user.rb [hunk 4]
```diff
@@ -1,8 +1,40 @@
...
+    reader = PDF::Reader.new(file[:tempfile].path)
+    save_next_line = false
+
+    reader.pages.each do |page|
+      page.text.each_line do |line|
+        # Remove blank lines
+        next if line.strip.empty?
+
+        if save_next_line
```
Comment (zbarnard1) [inline app/models/user.rb:20]: Without seeing an example of this PDF it's tough to say for sure, but this seems really complicated. There must be some data interspersed with the names, right? So we're basically using information from the header to determine if we save the next line, which should be a name?

If so, then we kindof have a Data Clump. https://refactoring.guru/smells/data-clumps

If we apply one of the treatments here and combine the "header"  and the "data" into one object, we could simplify this logic a bit. That will make it easier to reason about and less error-prone as a result.

File: app/app.rb (hunk 1 above)
Comment (ittullos) [reply app/app.rb:40]: Gotcha, so I'm pretty new to correctly naming routes. I've really only used this nested structure when using params. It definitely makes alot more sense though because we are creating the users residents. Do most request names refer to the class they are utilizing? Does that mean I should change my get /home route to get /verse?

File: app/models/user.rb (hunk 2 above)
Comment (ittullos) [reply app/models/user.rb:11]: Okay awesome! Yeah this did feel weird putting it in User. Makes alot more sense that way! Thanks!

File: spec/requests/app_spec.rb (hunk 3 above)
Comment (ittullos) [reply spec/requests/app_spec.rb:50]: Oh yikes did I miss the PDF file? whoops

Comment (ittullos) [reply spec/requests/app_spec.rb:50]: It doesn't show it here as a new file and is greyed out in the file system. When I click on it in VScode it's jibberish. Could this be because of encoding maybe?

File: app/models/user.rb (hunk 4 above)
Comment (ittullos) [reply app/models/user.rb:20]: Yeah there's probably a much better way to do this. The names are all on their own line. Below each line is link to pray, care, share, etc. So I'm basically looking for those link names and taking the line after it (after blank lines are removed). And then I have to look for the first line as well so I can get the first name. Maybe we could spend a little time on this over the weekend if you still have some time to hang.

File: app/app.rb (hunk 1 above)
Comment (zbarnard1) [reply app/app.rb:40]: This is a dense topic, but generally for something to be REST-ful, you'd name your routes according to a certain convention. Usually APIs fall short of fully adhering to REST, but it's a good guideline.
For this app, I don't think we need to go full on REST, so I don't think we need to change /home, since more stuff is probably going to get added to that response, but I am making assumptions here.

For what it's worth, my comment was more of a question to get you thinking about this. We can go with whatever name makes sense, but we should try to be consistent so that we don't have to think too hard.

For reference on REST: https://restfulapi.net

File: spec/requests/app_spec.rb (hunk 3 above)
Comment (zbarnard1) [reply spec/requests/app_spec.rb:50]: Yeah PDFs are encoded in a strange way. They have a certain header at the beginning that tells the thing reading it that it's a PDF, and usually you CAN open it in VSCode but it won't be fully readable. Treat it like you'd treat an image that you're committing to source control.

File: app/models/user.rb (hunk 4 above)
Comment (zbarnard1) [reply app/models/user.rb:20]: Ok, so maybe we can filter out the links and just process a list of names one by one?

**Comment by ittullos** [comment]
Okay so I created a new service class to parse the PDF file and create the User's Residents. The tests are a bit bare boned at the moment. Would love to know what else I should test. I also refactored the the pdf parser to only eliminate any lines that contain a keyword or are empty instead of taking lines in relation to the text in the other lines. The only thing is I had to eliminate addresses based off of whether or not there was a number in the line. I looked it up online and some people say it is possible to have an address without a number but I guess that is an edge case. Worst case someone will pray for an address instead of a person. I also changed the app and app_spec. My test is failing but I'm not sure about the setup and what exactly I should be testing.

File: app/services/pdf_parser.rb [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+  end
+
+  def load_residents
+    reader = PDF::Reader.new(@file[:tempfile].path)
+    skip_words = ['information', 'page', 'Unknown','disciple', 'Neighborhood', 'Powered', 'credit card']
+
+    reader.pages.each do |page|
+      page.text.each_line do |line|
+        # Remove blank lines or lines containing skip words
+        next if line.strip.empty? || skip_words.any? { |word| line.include?(word) }
```
Comment (zbarnard1) [inline app/services/pdf_parser.rb:16]: This is SO much better! Great progress!

File: spec/services/pdf_parser_spec.rb [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+RSpec.describe PdfParser do
+  context 'when the PDF file is valid' do
+    let(:pdf_file) { Rack::Test::UploadedFile.new('spec/fixtures/sample.pdf', 'application/pdf') }
+    let!(:user) { authenticated }
+
+    it 'processes the uploaded PDF file' do
+      resident_count = Resident.count
+      PdfParser.new({tempfile: pdf_file}, user).load_residents
+
+      expect(Resident.count).to eq(resident_count + 23)
```
Comment (zbarnard1) [inline spec/services/pdf_parser_spec.rb:16]: You could also use something like `expect { do_the_thing }.to change(Resident, :count).by(23)` here, which I think is slightly more expressive.

File: spec/services/pdf_parser_spec.rb [hunk 7]
```diff
@@ -0,0 +1,19 @@
+require 'spec_helper'
+require_relative '../../app/services/pdf_parser'
+require './app/models/user'
+
+include AuthenticationSpecHelpers
+
+RSpec.describe PdfParser do
+  context 'when the PDF file is valid' do
+    let(:pdf_file) { Rack::Test::UploadedFile.new('spec/fixtures/sample.pdf', 'application/pdf') }
+    let!(:user) { authenticated }
```
Comment (zbarnard1) [inline spec/services/pdf_parser_spec.rb:10]: We don't need them to be authenticated for these tests, these should just be focused on the PdfParser. As long as the User is in the DB (via FactoryBot.create(:user) then it should work.

File: app/services/pdf_parser.rb (hunk 5 above)
Comment (zbarnard1) [reply app/services/pdf_parser.rb:16]: Here's a more complete example of what I suggested in my text.
Pull the logic for parsing the PDF into a method and just have that return an array of strings. The method just reads the @file and returns an array of all the lines, then you can filter it and create Residents as needed.
In the test, you can do:
```ruby
allow(subject).to receive(:parse_pdf).and return(['foobar', 'Tommy Jones', 'information', 'Elon Musk', 'etc.'])
expect { subject.load_residents }.to change(Resident, :count).by(2)
```

**Comment by ittullos** [comment]
Made some changes to my pdf parser spec. I feel like I'm starting to understand mocking and stubbing but still a little confused. 

1) Right now my test is failing because the Resident count has not changed by 2. How does the count get increased with a stubbed method? (not sure if I'm using the right lingo here) Also am I even testing the right thing?

2) I still haven't changed the let!(:user) { authenticated } line yet. Do I even need to have a :user here? if so can I just exchange the authenticated line with User.new(...)?

3) Does it make sense to do this setup for every test or should I just put it into a before block?

File: spec/services/pdf_parser_spec.rb [hunk 8]
```diff
@@ -5,15 +5,30 @@
 include AuthenticationSpecHelpers
 
 RSpec.describe PdfParser do
-  context 'when the PDF file is valid' do
-    let(:pdf_file) { Rack::Test::UploadedFile.new('spec/fixtures/sample.pdf', 'application/pdf') }
-    let!(:user) { authenticated }
+  let(:pdf_file) { Rack::Test::UploadedFile.new('spec/fixtures/sample.pdf', 'application/pdf') }
+  let!(:user) { authenticated }
+  let(:pdf_parser) { double(PdfParser) }
```
Comment (zbarnard1) [inline spec/services/pdf_parser_spec.rb:10]: We shouldn't have a double for the object under test. You can use a double to represent something that you don't have control over, or is outside the scope of the test, but we need to be testing the real PdfParser.

**Comment by ittullos** [comment]
Made the changes you suggested in the app file and moved the pdf parser into a module. Also set up a super basic test for reading the pdf. My test for when the file is invalid is failing. This has to be something with the way I have set the test up. I'm getting a no method error instead of an error about the file. Feels like one of those helper methods with a formatted response might be helpful but not sure where to put it. Also the tests for the parser itself are incomplete and would love some suggestions on what else I should test.

**Comment by ittullos** [comment]
Fixed my app_spec and also wrote all the tests for my resident_list_spec. Sorry the commit history is a little messy. Probably going to have to reorganize everything at the end. While I was writing the tests for resident_list I found several mistakes I had made and fixed those. One big question I have is am I setting up my tests correctly in resident_list_spec. I'm importing some pdf files that are in my fixtures and then running tests on them. I also saw that there is a gem called "Prawn" that will allow me to create a pdf file on the fly and delete it when I'm done. That kinda seemed like overkill but def down to change it if you think that works better. All tests are passing currently.

File: app/models/user.rb [hunk 9]
```diff
@@ -1,5 +1,7 @@
 # frozen_string_literal: true
 
+require 'pdf-reader'
```
Comment (zbarnard1) [inline app/models/user.rb:3]: We don't need this on the User model, only in the class that actually reads the PDF.

File: spec/services/pdf_parser_spec.rb (hunk 6 above)
Comment (zbarnard1) [reply spec/services/pdf_parser_spec.rb]: I think this comment still stands. You can do basically what you're doing here with just my suggestion. the `change` matcher checks the count before you start and then at the end.

File: app/services/resident_list.rb [hunk 10]
```diff
@@ -0,0 +1,38 @@
+require 'pdf-reader'
+
+class InvalidFileFormatError < StandardError; end
```
Comment (zbarnard1) [inline app/services/resident_list.rb:3]: This should be within the ResidentList::PDF class. It's going to be namespaced with those names, so you may need to update a test where you reference it. It'll be called `ResidentList::PDF::InvalidFormatError`, which is a little more descriptive if it's ever logged.

File: app/services/resident_list.rb [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+
+module ResidentList
+  class PDF
+    def initialize(file)
+      @file = file
+    end
+
+    def load_residents
+      # Guard clause to check if @file is a PDF
```
Comment (zbarnard1) [inline app/services/resident_list.rb:14]: You can remove these comments for the guard clauses.  You should have an empty line after guard clauses, which I think you generally do but just wanted to say so.

File: app/app.rb [hunk 12]
```diff
@@ -38,6 +38,12 @@
 end
 
 post '/user/residents' do
+  # Guard clause to check if the file parameter is present
```
Comment (zbarnard1) [inline app/app.rb:41]: The code reads well enough that we don't need this comment, either.

**Review by zbarnard1 — CHANGES_REQUESTED** [review]
It's cool to see your progression as a developer even through these 4 commits, there's some good stuff here!
I would go ahead and squash these into 1 commit using a rebase (rebase onto `restore-sequel`) and then do a `git reset HEAD~1` to undo the squash commit. Then you can commit things in logical chunks. I generally start with the lowest level things first (like a new service class and its tests), then implement the new class in whatever route it's used in. Dependencies can go either in their own commit or where they are first used.

File: spec/requests/app_spec.rb [hunk 13]
```diff
@@ -46,6 +48,41 @@
     end
   end
 
+  describe 'POST /user/residents' do
+    context 'when the file is valid' do
+      before do
+        allow(ResidentList::PDF).to receive(:new).and_return(
```
Comment (zbarnard1) [inline spec/requests/app_spec.rb:54]: Request specs are more of an "integration" spec, so we should try to use the real objects anytime we can. Since we do have a good PDF file and it's used here in the tests already, let's test against that instead of stubbing out `#load_residents`. I would remove this whole before block and test using a real instance of the ResidentList::PDF class.

**Review by zbarnard1 — CHANGES_REQUESTED** [review]
Nice work on the rebase, this is much easier to review than before! It's very very close, but I have 2(maybe 3) main issues:
You will need to rebase it to resolve the merge conflicts in the Gemfile.lock. I have a hunch that your local restore-sequel branch is out of date, so you'll need to pull that first before you rebase.
Let's use the real ResidentList::PDF class in the request specs rather than a double since we do own that code and we have a good PDF file to use.
Check that the ResidentList::PDF class is properly required (possibly in app.rb rather than the spec file) so that it'll work when it's deployed. I'm not confident this is broken but I suspect it might be, so it's worth confirming.

File: spec/requests/app_spec.rb [hunk 14]
```diff
@@ -1,6 +1,7 @@
 # frozen_string_literal: true
 
 require 'spec_helper'
+require_relative '../../app/services/resident_list'
```
[the remark this answers is not in the document]
Comment (ittullos) [reply spec/requests/app_spec.rb:4]: Yep!! Good call!

File: spec/services/resident_list_spec.rb [hunk 15]
```diff
@@ -0,0 +1,45 @@
+require 'spec_helper'
+require_relative '../../app/services/resident_list'
+
+include AuthenticationSpecHelpers
```
Comment (zbarnard1) [inline spec/services/resident_list_spec.rb:4]: This include is in the wrong spot, this needs to be inside the RSpec.describe block if you are going to use it.
However, we don't even need this module here. This is only if we need to test the API and need to include a JWT in the headers for an authenticated user.

File: spec/requests/app_spec.rb [hunk 16]
```diff
@@ -46,6 +48,35 @@
     end
   end
 
+  describe 'POST /user/residents' do
+    context 'when the file is valid' do
+      it 'creates a resident for each name in the file' do
+        post '/user/residents', { file: pdf_file }, headers
+
+        expect(last_response.status).to eq(200)
+        expect(Resident.count).to eq(23)
```
Comment (zbarnard1) [inline spec/requests/app_spec.rb:57]: ⛏️ Usually we'd see this as `expect { post ... }.to change(Resident, :count).by(23)`. We're using DatabaseCleaner so I don't think we could actually get a false positive in practice, but it's a little bit more "correct" to do it this way.

File: spec/services/resident_list_spec.rb [hunk 17]
```diff
@@ -0,0 +1,45 @@
...
+    end
+
+    it 'skips over duplicate names' do
+      residents = ResidentList::PDF.new(duplicate_name_sample).load_residents
+      expect(residents.count).to eq(3)
+    end
+
+  end
+
+  context 'when the file is not a pdf' do
```
Comment (zbarnard1) [inline spec/services/resident_list_spec.rb:28]: ⛏️ These three tests have some very long lines, it helps readability if you split them so there's a maximum of like 80 characters. Usually you can put the two arguments to `raise_error` on a new line, or you can make your block a multiline block with a `do end`, whichever you think is best for the situation.

**Review by zbarnard1 — APPROVED** [review]
Looks good! I have some nitpicks but nothing major. There's probably some room for optimization but since we're only dealing with 
I'm approving this with the assumption that you'll remove that extra `include` that breaks those specs before you merge. Thanks for sticking with this and seeing it through, I know I left a lot of comments along the way but I'm glad to see it coming together!

File: spec/services/resident_list_spec.rb (hunk 15 above)
Comment (ittullos) [reply spec/services/resident_list_spec.rb:4]: Gotcha! 🫡
